gematria/datasets/exegesis_benchmark.cc

```diff
@@ -0,0 +1,325 @@
+          "definition is not a JSON object");
+
+    std::optional<StringRef> MemoryDefinitionName =
+        MemoryDefinitionObject->getString("Name");
+    std::optional<int64_t> MemoryDefinitionSize =
+        MemoryDefinitionObject->getInteger("Size");
+
+    std::optional<int64_t> MemoryDefinitionHexValue =
+        MemoryDefinitionObject->getInteger("Value");
+
+    if (!MemoryDefinitionName.has_value() ||
+        !MemoryDefinitionSize.has_value() ||
+        !MemoryDefinitionHexValue.has_value())
+      return llvm::make_error<StringError>(
+          errc::invalid_argument,
+          "Malformed memory definition: no size, name, or value");
+
+    MemoryValue MemVal;
+    MemVal.Value = APInt(32, *MemoryDefinitionHexValue);
+    MemVal.Index = MemoryDefinitionIndex;
+    MemVal.SizeBytes = *MemoryDefinitionSize;
+
+    BenchCode.Key.MemoryValues[MemoryDefinitionName->str()] = MemVal;
+
+    ++MemoryDefinitionIndex;
+  }
+
+  const json::Array *MemoryMappings = BasicBlockJSON.getArray("MemoryMappings");
+
+  if (!MemoryMappings)
+    return llvm::make_error<StringError>(
+        errc::invalid_argument,
+        "Malformed basic block: no memory mappings array");
+
+  for (const auto &MemoryMappingValue : *MemoryMappings) {
+    const json::Object *MemoryMappingObject = MemoryMappingValue.getAsObject();
+
+    if (!MemoryMappingObject)
+      return llvm::make_error<StringError>(
+          errc::invalid_argument,
+          "Malformed memory mapping: memory mapping is not a JSON object");
+
+    std::optional<StringRef> MemoryMappingDefinitionName =
+        MemoryMappingObject->getString("Value");
+    std::optional<uintptr_t> MemoryMappingAddress =
+        MemoryMappingObject->getInteger("Address");
+
+    if (!MemoryMappingDefinitionName.has_value() ||
+        !MemoryMappingAddress.has_value())
+      return llvm::make_error<StringError>(
+          errc::invalid_argument,
+          "Malformed memory mapping: no name or address");
+
+    MemoryMapping MemMap;
+    MemMap.Address = *MemoryMappingAddress;
+    MemMap.MemoryValueName = MemoryMappingDefinitionName->str();
+    BenchCode.Key.MemoryMappings.push_back(MemMap);
+  }
+
+  return BenchCode;
+}
+
+Expected<unsigned> benchmarkBasicBlock(const BenchmarkCode &BenchCode,
+                                       const BenchmarkRunner &BenchRunner,
+                                       const LLVMState &State) {
+  std::unique_ptr<const SnippetRepetitor> SnipRepetitor =
+      SnippetRepetitor::Create(Benchmark::RepetitionModeE::MiddleHalfLoop,
+                               State, BenchCode.Key.LoopRegister);
+
+  SmallVector<Benchmark, 2> AllResults;
+
+  auto RC1 =
+      BenchRunner.getRunnableConfiguration(BenchCode, 5000, 0, *SnipRepetitor);
+  if (!RC1) return RC1.takeError();
+  auto RC2 =
+      BenchRunner.getRunnableConfiguration(BenchCode, 10000, 0, *SnipRepetitor);
+  if (!RC2) return RC2.takeError();
+
+  std::pair<Error, Benchmark> BenchmarkResult1OrErr =
```

Comment:
The variables were named `BenchmarkResult1<suffix>` and `BenchmarkResult2<suffix>` since we need two to run the comparison (rather than 10, which is what the capital O makes it look like). I've switched to A/B to make things more clear.